keras_cv/models/__internal__/unet.py

```diff
@@ -0,0 +1,172 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from tensorflow.keras import initializers
+from tensorflow.keras import layers
+from tensorflow.keras import regularizers
+
+
+def Block(filters, downsample):
+    def apply(x):
+        input_depth = x.shape.as_list()[-1]
+        stride = 2 if downsample else 1
+
+        residual = x
+
+        x = layers.Conv2D(
+            filters,
+            3,
+            stride,
+            padding="same",
+            kernel_initializer=initializers.VarianceScaling(),
+            kernel_regularizer=regularizers.L2(l2=1e-4),
+        )(x)
+        x = layers.BatchNormalization(
+            synchronized=True,
+            beta_regularizer=regularizers.L2(l2=1e-8),
+            gamma_regularizer=regularizers.L2(l2=1e-8),
+        )(x)
+        x = layers.ReLU()(x)
+
+        x = layers.Conv2D(
+            filters,
+            3,
+            1,
+            padding="same",
+            kernel_initializer=initializers.VarianceScaling(),
+            kernel_regularizer=regularizers.L2(l2=1e-4),
+        )(x)
+        x = layers.BatchNormalization(
+            synchronized=True,
+            beta_regularizer=regularizers.L2(l2=1e-8),
+            gamma_regularizer=regularizers.L2(l2=1e-8),
+        )(x)
+
+        if downsample:
+            residual = layers.MaxPool2D(pool_size=2, strides=2, padding="SAME")(
+                residual
+            )
+
+        if input_depth != filters:
+            residual = layers.Conv2D(
+                filters,
+                1,
+                1,
+                padding="same",
+                kernel_initializer=initializers.VarianceScaling(),
+                kernel_regularizer=regularizers.L2(l2=1e-4),
+            )(residual)
+
+        x = x + residual
+        x = layers.ReLU()(x)
+
+        return x
+
+    return apply
+
+
+def SkipBlock(filters):
+    def apply(x):
+        x = layers.Conv2D(
+            filters,
+            1,
+            1,
+            kernel_initializer=initializers.VarianceScaling(),
+            kernel_regularizer=regularizers.L2(l2=1e-4),
+        )(x)
+        x = layers.BatchNormalization(
+            synchronized=True,
+            beta_regularizer=regularizers.L2(l2=1e-8),
+            gamma_regularizer=regularizers.L2(l2=1e-8),
+        )(x)
+        x = layers.ReLU()(x)
+
+        return x
+
+    return apply
+
+
+def DownSampleBlock(filters, num_blocks):
+    def apply(x):
+        x = Block(filters, downsample=True)(x)
+
+        for _ in range(num_blocks - 1):
+            x = Block(filters, downsample=False)(x)
+
+        return x
+
+    return apply
+
+
+def UpSampleBlock(filters):
+    def apply(x, lateral_input):
+        x = layers.Conv2DTranspose(
```

Comment:
Is the UNET paper using UpSampling2D or Conv2DTranspose? Can this be configurable?

Reply:
Bilinear is slower than ConvTranspose on TPU. We could support bilinear up sampling once we verify the unet performance.

Reply:
The idea here is that this UpSampleBlock (as well as the DownSampleBlock) are specific to this UNet.

When this is a public API, we could example call `UNet(...., down_block=MyCustomDownBlock)`
In that future, these particular up/down blocks would be available as e.g. a CenterPillarDownBlock

Reply:
Sounds good!

Reply:
Ah sounds good.